Provided a dataset named messy_data.csv. The task is to clean this dataset and ensure it is ready for analysis. 

## 1) Load the data

In [401]:
import pandas as pd
given_data=pd.read_csv('messy_data.csv')

First, Pandas library is imported and the given data csv file called messy_data.csv is loaded into a variable 'data'. The variable 'data' will hold a Pandas DataFrame.

In [402]:
data = given_data.copy()

## 2)Inspect the data

In [403]:
# return first five rows
data.head()

Unnamed: 0                                    ID                     Name  \
0           0  1e407ff9-6255-489d-a0de-34135d4f74bd            Hunter Thomas   
1           1  379f55b8-87d5-4739-a146-7400b78c24d1             Jeremy Irwin   
2           2  18261368-dfa1-47f0-afc6-bddf45926b07  Jennifer Hammondquickly   
3           3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5          Sydney Taylorso   
4           4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90                Julia Lee   

    Age                    Email   Join Date    Salary   Department  
0  25.0       xlopez@hotmail.com         NaN   88552.0        Sales  
1  90.0          Jillian Jenkins  2022-07-07  139227.0          NaN  
2  66.0          jscottgreen.biz  2023-11-21   65550.0  Engineering  
3  39.0       luke56gonzalez.com  2021-11-05  139932.0     SupportJ  
4  71.0  figueroakayla@yahoo.com         NaN  143456.0    Marketing

The first five rows of the DataFrame are shown. It is understood that the table includes the ID of a candidate, their age, their email, their join date, the salary they receive, and the department to which they belong. Some of the columns contain NaN values; these values need to be handled.

In [404]:
# returning last five rows
data.tail()

Unnamed: 0                                    ID                Name  \
10995        6523  07c223be-03e6-4f70-a2b5-86df778cc61a                 NaN   
10996        9785  da8a6bbc-5026-4630-848d-f64e80dac56c    Steven Armstrong   
10997        7826  ed19c966-d6d8-4047-b410-b6e595a39340  Stephanie Riossell   
10998        7648  783b36b4-d09f-46c9-8a52-7ff96b80863e      Bonnie Benitez   
10999        7107  fc25a38a-5747-46eb-b6d3-7173f8255809      Caroline Ochoa   

        Age                      Email   Join Date         Salary Department  
10995   NaN                        NaN         NaN            NaN        NaN  
10996  38.0           molly89gmail.com  2021-06-24            NaN      Sales  
10997   NaN  robert96@pollard-frye.com  15/08/2006  122609.594149         HR  
10998  37.0         roypark@warren.net  2020-10-09  147322.005171    Support  
10999  53.0          cdavis@hodges.com  2023-08-10  149224.000000    Support

Similarly, the last five rows are shown, and NaN values also seem to be present here.

In [405]:
# overview of the data including datatypes
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11000 entries, 0 to 10999
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  11000 non-null  int64  
 1   ID          11000 non-null  object 
 2   Name        8667 non-null   object 
 3   Age         9253 non-null   float64
 4   Email       9731 non-null   object 
 5   Join Date   8808 non-null   object 
 6   Salary      8761 non-null   float64
 7   Department  8745 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 687.6+ KB


Here, the column names, count of non-null values, and their data types are shown. Upon observation, the Age and Salary columns contain float values, while the Join Date column is of object type. Therefore, they need to be converted into integers and datetime datatypes, respectively.

In [406]:
# providing a statistical summary 
data.describe()

Unnamed: 0          Age         Salary
count  11000.000000  9253.000000    8761.000000
mean    5012.947818    54.162650   89886.585012
std     2884.739158    21.072919   34896.320117
min        0.000000    18.000000   24655.136613
25%     2509.750000    36.000000   59723.844874
50%     5024.500000    54.000000   89241.000000
75%     7510.250000    72.000000  119491.000000
max     9999.000000    90.000000  176156.206747

This function, .describe(), provides a statistical summary of the entire DataFrame. Since Age and Salary are integers (or floats in this case), only their statistical values can be represented. Considering the non-null rows of these columns, the count, mean, standard deviation, maximum value, minimum value, first quartile(25%), second quartile(50%), and third quartile(75%) values are shown.

In [407]:
# returning number of rows and columns
data.shape

(11000, 8)

The Data Frame consists of 11000 rows and 8 columns. Some of the rows/columns will be removed in order to keep the data clean.

In [408]:
# count the missing values
data.isnull().sum()

Unnamed: 0       0
ID               0
Name          2333
Age           1747
Email         1269
Join Date     2192
Salary        2239
Department    2255
dtype: int64

isnull().sum()' counts the null, missing, or NaN values. These values can be substituted by the mean, median, or mode of their respective columns.

In [409]:
# count of duplicate rows
dup = data.duplicated().sum()
dup

np.int64(291)

The dataset has 291 duplicate rows. These rows should be removed, as they may create data inconsistency.

## 3) Record QA Issues: 

**Unnamed**:    
No null values. No inconsistency. No data issues in this column.    
**ID**:     
No null values.   
Duplicates in the ID column     
**Name**:    
2333 null values present.    
Some of them having unwanted words along with their last name.    
Some of them having honorifics in the beginning of the name.    
Some of them having professional or academic degrees long with their last name.   
**Age**   
Age is of float datatype.   
1747 null values present     
**Email**  
1269 null values present.
Some of them are invalid emails like without @ or just a name with first and last name.    
Some of them are non-professional emails like yahoo.com, gmail.com, hotmail.com.   
Also, ending with .net,.biz,.io.   
**Join Date**   
2192 null values present.   
Some of them are not in the exact format i.e(YYYY-MM-DD)   
**Salary**   
Salary is of float datatype.   
2239 null values present.   
**Department**     
2255 null values present.   
Some of them have umwanted letters followed by the Department name like Supportj, Marketingl, HRS, Salese, engineeringZ

## 4) Handle missing values

In [410]:
data

Unnamed: 0                                    ID  \
0               0  1e407ff9-6255-489d-a0de-34135d4f74bd   
1               1  379f55b8-87d5-4739-a146-7400b78c24d1   
2               2  18261368-dfa1-47f0-afc6-bddf45926b07   
3               3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5   
4               4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90   
...           ...                                   ...   
10995        6523  07c223be-03e6-4f70-a2b5-86df778cc61a   
10996        9785  da8a6bbc-5026-4630-848d-f64e80dac56c   
10997        7826  ed19c966-d6d8-4047-b410-b6e595a39340   
10998        7648  783b36b4-d09f-46c9-8a52-7ff96b80863e   
10999        7107  fc25a38a-5747-46eb-b6d3-7173f8255809   

                          Name   Age                      Email   Join Date  \
0                Hunter Thomas  25.0         xlopez@hotmail.com         NaN   
1                 Jeremy Irwin  90.0            Jillian Jenkins  2022-07-07   
2      Jennifer Hammondquickly  66.0            jscottgreen.biz  2023-11-21   
3              Sydney Taylorso  39.0         luke56gonzalez.com  2021-11-05   
4                    Julia Lee  71.0    figueroakayla@yahoo.com         NaN   
...                        ...   ...                        ...         ...   
10995                      NaN   NaN                        NaN         NaN   
10996         Steven Armstrong  38.0           molly89gmail.com  2021-06-24   
10997       Stephanie Riossell   NaN  robert96@pollard-frye.com  15/08/2006   
10998           Bonnie Benitez  37.0         roypark@warren.net  2020-10-09   
10999           Caroline Ochoa  53.0          cdavis@hodges.com  2023-08-10   

              Salary   Department  
0       88552.000000        Sales  
1      139227.000000          NaN  
2       65550.000000  Engineering  
3      139932.000000     SupportJ  
4      143456.000000    Marketing  
...              ...          ...  
10995            NaN          NaN  
10996            NaN        Sales  
10997  122609.594149           HR  
10998  147322.005171      Support  
10999  149224.000000      Support  

[11000 rows x 8 columns]

The data is shown again for a deeper understanding. One row contains 6 NaN values, and there may be more such rows. These rows should be removed.

In [411]:
# Consider to remove rows with more than 3 NaN values
data = data[data.isna().sum(axis=1) < 4]
data

Unnamed: 0                                    ID  \
0               0  1e407ff9-6255-489d-a0de-34135d4f74bd   
1               1  379f55b8-87d5-4739-a146-7400b78c24d1   
2               2  18261368-dfa1-47f0-afc6-bddf45926b07   
3               3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5   
4               4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90   
...           ...                                   ...   
10994        3978  6714ff24-3600-4c6e-8eed-5b4224bd9e44   
10996        9785  da8a6bbc-5026-4630-848d-f64e80dac56c   
10997        7826  ed19c966-d6d8-4047-b410-b6e595a39340   
10998        7648  783b36b4-d09f-46c9-8a52-7ff96b80863e   
10999        7107  fc25a38a-5747-46eb-b6d3-7173f8255809   

                          Name   Age                      Email   Join Date  \
0                Hunter Thomas  25.0         xlopez@hotmail.com         NaN   
1                 Jeremy Irwin  90.0            Jillian Jenkins  2022-07-07   
2      Jennifer Hammondquickly  66.0            jscottgreen.biz  2023-11-21   
3              Sydney Taylorso  39.0         luke56gonzalez.com  2021-11-05   
4                    Julia Lee  71.0    figueroakayla@yahoo.com         NaN   
...                        ...   ...                        ...         ...   
10994              Karen Wells  29.0    cheryllewis@hotmail.com         NaN   
10996         Steven Armstrong  38.0           molly89gmail.com  2021-06-24   
10997       Stephanie Riossell   NaN  robert96@pollard-frye.com  15/08/2006   
10998           Bonnie Benitez  37.0         roypark@warren.net  2020-10-09   
10999           Caroline Ochoa  53.0          cdavis@hodges.com  2023-08-10   

              Salary   Department  
0       88552.000000        Sales  
1      139227.000000          NaN  
2       65550.000000  Engineering  
3      139932.000000     SupportJ  
4      143456.000000    Marketing  
...              ...          ...  
10994  147555.000000  Engineering  
10996            NaN        Sales  
10997  122609.594149           HR  
10998  147322.005171      Support  
10999  149224.000000      Support  

[9725 rows x 8 columns]

In [412]:
# count the missing values
data.isnull().sum()

Unnamed: 0       0
ID               0
Name          1058
Age            476
Email            0
Join Date      917
Salary         965
Department     981
dtype: int64

In [413]:
# Handling NaN values with mean or mode

data.loc[:,'Name'] = data['Name'].fillna(data['Name'].mode()[0])
data.loc[:,'Email'] = data['Email'].fillna(data['Email'].mode()[0])
data.loc[:,'Salary'] = data['Salary'].fillna(data['Salary'].mean())
data.loc[:,'Department'] = data['Department'].fillna(data['Department'].mode()[0])
data.loc[:,'Join Date'] = data['Join Date'].fillna(data['Join Date'].mode()[0])
data

Unnamed: 0                                    ID  \
0               0  1e407ff9-6255-489d-a0de-34135d4f74bd   
1               1  379f55b8-87d5-4739-a146-7400b78c24d1   
2               2  18261368-dfa1-47f0-afc6-bddf45926b07   
3               3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5   
4               4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90   
...           ...                                   ...   
10994        3978  6714ff24-3600-4c6e-8eed-5b4224bd9e44   
10996        9785  da8a6bbc-5026-4630-848d-f64e80dac56c   
10997        7826  ed19c966-d6d8-4047-b410-b6e595a39340   
10998        7648  783b36b4-d09f-46c9-8a52-7ff96b80863e   
10999        7107  fc25a38a-5747-46eb-b6d3-7173f8255809   

                          Name   Age                      Email   Join Date  \
0                Hunter Thomas  25.0         xlopez@hotmail.com  2022-03-31   
1                 Jeremy Irwin  90.0            Jillian Jenkins  2022-07-07   
2      Jennifer Hammondquickly  66.0            jscottgreen.biz  2023-11-21   
3              Sydney Taylorso  39.0         luke56gonzalez.com  2021-11-05   
4                    Julia Lee  71.0    figueroakayla@yahoo.com  2022-03-31   
...                        ...   ...                        ...         ...   
10994              Karen Wells  29.0    cheryllewis@hotmail.com  2022-03-31   
10996         Steven Armstrong  38.0           molly89gmail.com  2021-06-24   
10997       Stephanie Riossell   NaN  robert96@pollard-frye.com  15/08/2006   
10998           Bonnie Benitez  37.0         roypark@warren.net  2020-10-09   
10999           Caroline Ochoa  53.0          cdavis@hodges.com  2023-08-10   

              Salary   Department  
0       88552.000000        Sales  
1      139227.000000      Support  
2       65550.000000  Engineering  
3      139932.000000     SupportJ  
4      143456.000000    Marketing  
...              ...          ...  
10994  147555.000000  Engineering  
10996   89883.332796        Sales  
10997  122609.594149           HR  
10998  147322.005171      Support  
10999  149224.000000      Support  

[9725 rows x 8 columns]

In [414]:
#age to integer
mean_age = data['Age'].mean()
data.loc[:, 'Age'] = data['Age'].fillna(mean_age)
data

Unnamed: 0                                    ID  \
0               0  1e407ff9-6255-489d-a0de-34135d4f74bd   
1               1  379f55b8-87d5-4739-a146-7400b78c24d1   
2               2  18261368-dfa1-47f0-afc6-bddf45926b07   
3               3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5   
4               4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90   
...           ...                                   ...   
10994        3978  6714ff24-3600-4c6e-8eed-5b4224bd9e44   
10996        9785  da8a6bbc-5026-4630-848d-f64e80dac56c   
10997        7826  ed19c966-d6d8-4047-b410-b6e595a39340   
10998        7648  783b36b4-d09f-46c9-8a52-7ff96b80863e   
10999        7107  fc25a38a-5747-46eb-b6d3-7173f8255809   

                          Name        Age                      Email  \
0                Hunter Thomas  25.000000         xlopez@hotmail.com   
1                 Jeremy Irwin  90.000000            Jillian Jenkins   
2      Jennifer Hammondquickly  66.000000            jscottgreen.biz   
3              Sydney Taylorso  39.000000         luke56gonzalez.com   
4                    Julia Lee  71.000000    figueroakayla@yahoo.com   
...                        ...        ...                        ...   
10994              Karen Wells  29.000000    cheryllewis@hotmail.com   
10996         Steven Armstrong  38.000000           molly89gmail.com   
10997       Stephanie Riossell  54.163477  robert96@pollard-frye.com   
10998           Bonnie Benitez  37.000000         roypark@warren.net   
10999           Caroline Ochoa  53.000000          cdavis@hodges.com   

        Join Date         Salary   Department  
0      2022-03-31   88552.000000        Sales  
1      2022-07-07  139227.000000      Support  
2      2023-11-21   65550.000000  Engineering  
3      2021-11-05  139932.000000     SupportJ  
4      2022-03-31  143456.000000    Marketing  
...           ...            ...          ...  
10994  2022-03-31  147555.000000  Engineering  
10996  2021-06-24   89883.332796        Sales  
10997  15/08/2006  122609.594149           HR  
10998  2020-10-09  147322.005171      Support  
10999  2023-08-10  149224.000000      Support  

[9725 rows x 8 columns]

In [415]:
# Convert to integer 
data['Age'] = data['Age'].astype(int)
data

C:\Users\itsme\AppData\Local\Temp\ipykernel_31228\2805976082.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['Age'] = data['Age'].astype(int)


Unnamed: 0                                    ID  \
0               0  1e407ff9-6255-489d-a0de-34135d4f74bd   
1               1  379f55b8-87d5-4739-a146-7400b78c24d1   
2               2  18261368-dfa1-47f0-afc6-bddf45926b07   
3               3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5   
4               4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90   
...           ...                                   ...   
10994        3978  6714ff24-3600-4c6e-8eed-5b4224bd9e44   
10996        9785  da8a6bbc-5026-4630-848d-f64e80dac56c   
10997        7826  ed19c966-d6d8-4047-b410-b6e595a39340   
10998        7648  783b36b4-d09f-46c9-8a52-7ff96b80863e   
10999        7107  fc25a38a-5747-46eb-b6d3-7173f8255809   

                          Name  Age                      Email   Join Date  \
0                Hunter Thomas   25         xlopez@hotmail.com  2022-03-31   
1                 Jeremy Irwin   90            Jillian Jenkins  2022-07-07   
2      Jennifer Hammondquickly   66            jscottgreen.biz  2023-11-21   
3              Sydney Taylorso   39         luke56gonzalez.com  2021-11-05   
4                    Julia Lee   71    figueroakayla@yahoo.com  2022-03-31   
...                        ...  ...                        ...         ...   
10994              Karen Wells   29    cheryllewis@hotmail.com  2022-03-31   
10996         Steven Armstrong   38           molly89gmail.com  2021-06-24   
10997       Stephanie Riossell   54  robert96@pollard-frye.com  15/08/2006   
10998           Bonnie Benitez   37         roypark@warren.net  2020-10-09   
10999           Caroline Ochoa   53          cdavis@hodges.com  2023-08-10   

              Salary   Department  
0       88552.000000        Sales  
1      139227.000000      Support  
2       65550.000000  Engineering  
3      139932.000000     SupportJ  
4      143456.000000    Marketing  
...              ...          ...  
10994  147555.000000  Engineering  
10996   89883.332796        Sales  
10997  122609.594149           HR  
10998  147322.005171      Support  
10999  149224.000000      Support  

[9725 rows x 8 columns]

In [416]:
data.isnull().sum()

Unnamed: 0    0
ID            0
Name          0
Age           0
Email         0
Join Date     0
Salary        0
Department    0
dtype: int64

In [417]:
data

Unnamed: 0                                    ID  \
0               0  1e407ff9-6255-489d-a0de-34135d4f74bd   
1               1  379f55b8-87d5-4739-a146-7400b78c24d1   
2               2  18261368-dfa1-47f0-afc6-bddf45926b07   
3               3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5   
4               4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90   
...           ...                                   ...   
10994        3978  6714ff24-3600-4c6e-8eed-5b4224bd9e44   
10996        9785  da8a6bbc-5026-4630-848d-f64e80dac56c   
10997        7826  ed19c966-d6d8-4047-b410-b6e595a39340   
10998        7648  783b36b4-d09f-46c9-8a52-7ff96b80863e   
10999        7107  fc25a38a-5747-46eb-b6d3-7173f8255809   

                          Name  Age                      Email   Join Date  \
0                Hunter Thomas   25         xlopez@hotmail.com  2022-03-31   
1                 Jeremy Irwin   90            Jillian Jenkins  2022-07-07   
2      Jennifer Hammondquickly   66            jscottgreen.biz  2023-11-21   
3              Sydney Taylorso   39         luke56gonzalez.com  2021-11-05   
4                    Julia Lee   71    figueroakayla@yahoo.com  2022-03-31   
...                        ...  ...                        ...         ...   
10994              Karen Wells   29    cheryllewis@hotmail.com  2022-03-31   
10996         Steven Armstrong   38           molly89gmail.com  2021-06-24   
10997       Stephanie Riossell   54  robert96@pollard-frye.com  15/08/2006   
10998           Bonnie Benitez   37         roypark@warren.net  2020-10-09   
10999           Caroline Ochoa   53          cdavis@hodges.com  2023-08-10   

              Salary   Department  
0       88552.000000        Sales  
1      139227.000000      Support  
2       65550.000000  Engineering  
3      139932.000000     SupportJ  
4      143456.000000    Marketing  
...              ...          ...  
10994  147555.000000  Engineering  
10996   89883.332796        Sales  
10997  122609.594149           HR  
10998  147322.005171      Support  
10999  149224.000000      Support  

[9725 rows x 8 columns]

In [418]:
data.isnull().sum()

Unnamed: 0    0
ID            0
Name          0
Age           0
Email         0
Join Date     0
Salary        0
Department    0
dtype: int64

## 5) Remove Duplicates

In [419]:
# remove duplicates present in the data
data=data.drop_duplicates()
data

Unnamed: 0                                    ID  \
0               0  1e407ff9-6255-489d-a0de-34135d4f74bd   
1               1  379f55b8-87d5-4739-a146-7400b78c24d1   
2               2  18261368-dfa1-47f0-afc6-bddf45926b07   
3               3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5   
4               4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90   
...           ...                                   ...   
10993        9510  796f8c0b-15a4-475c-bb02-ac892df0c746   
10994        3978  6714ff24-3600-4c6e-8eed-5b4224bd9e44   
10996        9785  da8a6bbc-5026-4630-848d-f64e80dac56c   
10997        7826  ed19c966-d6d8-4047-b410-b6e595a39340   
10998        7648  783b36b4-d09f-46c9-8a52-7ff96b80863e   

                          Name  Age                      Email   Join Date  \
0                Hunter Thomas   25         xlopez@hotmail.com  2022-03-31   
1                 Jeremy Irwin   90            Jillian Jenkins  2022-07-07   
2      Jennifer Hammondquickly   66            jscottgreen.biz  2023-11-21   
3              Sydney Taylorso   39         luke56gonzalez.com  2021-11-05   
4                    Julia Lee   71    figueroakayla@yahoo.com  2022-03-31   
...                        ...  ...                        ...         ...   
10993      Shelley Hopkinslive   79             Michael Walker  25/12/1992   
10994              Karen Wells   29    cheryllewis@hotmail.com  2022-03-31   
10996         Steven Armstrong   38           molly89gmail.com  2021-06-24   
10997       Stephanie Riossell   54  robert96@pollard-frye.com  15/08/2006   
10998           Bonnie Benitez   37         roypark@warren.net  2020-10-09   

              Salary   Department  
0       88552.000000        Sales  
1      139227.000000      Support  
2       65550.000000  Engineering  
3      139932.000000     SupportJ  
4      143456.000000    Marketing  
...              ...          ...  
10993   47809.000000           HR  
10994  147555.000000  Engineering  
10996   89883.332796        Sales  
10997  122609.594149           HR  
10998  147322.005171      Support  

[9611 rows x 8 columns]

In [420]:
# remove rows with duplicated ID
data = data.drop_duplicates(subset=['ID'], keep='first')
data

Unnamed: 0                                    ID  \
0              0  1e407ff9-6255-489d-a0de-34135d4f74bd   
1              1  379f55b8-87d5-4739-a146-7400b78c24d1   
2              2  18261368-dfa1-47f0-afc6-bddf45926b07   
3              3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5   
4              4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90   
...          ...                                   ...   
9995        9995  610210a0-e344-4b79-95ce-ff6583d7d8ce   
9996        9996  eb66531c-901a-4da7-bc60-a4951886b5a8   
9997        9997  86a83f25-224d-498d-9901-c2e4a49ffbd2   
9998        9998  9cbba425-6af2-455f-8566-62402de79661   
9999        9999  ae6eee4d-ca8f-4932-96f1-487230875968   

                         Name  Age                              Email  \
0               Hunter Thomas   25                 xlopez@hotmail.com   
1                Jeremy Irwin   90                    Jillian Jenkins   
2     Jennifer Hammondquickly   66                    jscottgreen.biz   
3             Sydney Taylorso   39                 luke56gonzalez.com   
4                   Julia Lee   71            figueroakayla@yahoo.com   
...                       ...  ...                                ...   
9995          David Lambalone   70             jeremybanks@rogers.biz   
9996              Karen Yoder   59  smithdaniel@williams-thompson.net   
9997        Anthony Rasmussen   57                nfranklin@yahoo.com   
9998             Nicole Moore   82                   lmyers@gmail.com   
9999           Steven Gilbert   18                   cjones@yahoo.com   

       Join Date    Salary    Department  
0     2022-03-31   88552.0         Sales  
1     2022-07-07  139227.0       Support  
2     2023-11-21   65550.0   Engineering  
3     2021-11-05  139932.0      SupportJ  
4     2022-03-31  143456.0     Marketing  
...          ...       ...           ...  
9995  2020-12-17   88472.0       Support  
9996  2023-12-12   71575.0       Support  
9997  05/12/1979  128456.0     Marketing  
9998  2020-10-28  128939.0  EngineeringV  
9999  2022-06-01  101447.0            HR  

[8902 rows x 8 columns]

## 6) Correct Email Formats: 

In [421]:
#removing space between the string
data.loc[:, 'Email'] = data['Email'].str.replace(' ', '')
# Convering Any uppercase letters to lower case
data.loc[:, 'Email'] = data['Email'].str.lower()
def correct(email):
    if '@' not in email and '.' not in email:      # email without dot(.) and without @
       email+='@datahut.com'                        # appending with the name present in th email column 
    if '@' not in email and '.' in email:           # email without @ and with dot(.)
       email= email[0:-4]+'@datahut.com'            # removing last 4 characters and replacing with @datahut.com
    return email
    
data.loc[:,'Email'] = data['Email'].apply(correct)
# replace emails ending @yahoo.com or @gmail.com or @hotmail.com to @datahut.com
data.loc[:,'Email'] = data['Email'].str.replace(r'@(yahoo|gmail|hotmail)\.com', '@datahut.com', regex=True)
data.loc[:,'Email'] = data['Email'].str[:-4] + '.com'
data.loc[:, 'Email'] = data['Email'].str.replace(' ', '')
data.loc[:,'Email'] = data['Email'].str.replace(r'gmail', '', regex=True)
data.loc[:,'Email'] = data['Email'].str.replace(r'yahoo', '', regex=True)
data.loc[:,'Email'] = data['Email'].str.replace(r'hotmail', '', regex=True)

data

Unnamed: 0                                    ID  \
0              0  1e407ff9-6255-489d-a0de-34135d4f74bd   
1              1  379f55b8-87d5-4739-a146-7400b78c24d1   
2              2  18261368-dfa1-47f0-afc6-bddf45926b07   
3              3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5   
4              4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90   
...          ...                                   ...   
9995        9995  610210a0-e344-4b79-95ce-ff6583d7d8ce   
9996        9996  eb66531c-901a-4da7-bc60-a4951886b5a8   
9997        9997  86a83f25-224d-498d-9901-c2e4a49ffbd2   
9998        9998  9cbba425-6af2-455f-8566-62402de79661   
9999        9999  ae6eee4d-ca8f-4932-96f1-487230875968   

                         Name  Age                              Email  \
0               Hunter Thomas   25                 xlopez@datahut.com   
1                Jeremy Irwin   90         jillianjenkins@datahut.com   
2     Jennifer Hammondquickly   66            jscottgreen@datahut.com   
3             Sydney Taylorso   39         luke56gonzalez@datahut.com   
4                   Julia Lee   71          figueroakayla@datahut.com   
...                       ...  ...                                ...   
9995          David Lambalone   70             jeremybanks@rogers.com   
9996              Karen Yoder   59  smithdaniel@williams-thompson.com   
9997        Anthony Rasmussen   57              nfranklin@datahut.com   
9998             Nicole Moore   82                 lmyers@datahut.com   
9999           Steven Gilbert   18                 cjones@datahut.com   

       Join Date    Salary    Department  
0     2022-03-31   88552.0         Sales  
1     2022-07-07  139227.0       Support  
2     2023-11-21   65550.0   Engineering  
3     2021-11-05  139932.0      SupportJ  
4     2022-03-31  143456.0     Marketing  
...          ...       ...           ...  
9995  2020-12-17   88472.0       Support  
9996  2023-12-12   71575.0       Support  
9997  05/12/1979  128456.0     Marketing  
9998  2020-10-28  128939.0  EngineeringV  
9999  2022-06-01  101447.0            HR  

[8902 rows x 8 columns]

## 7. Clean Name Fields: 


In [422]:
import re
def remove_extra(data, name_column):
    def clean_name(name):
        # Remove titles Mr. , Mrs. ,Ms. ,Dr. ,Jr.
        name = re.sub(r'\b(Mr\.|Mrs\.|Ms\.|Dr\.|Jr\.)\s*', '', name)
        # Remove MD, DVM, DDS, PhD, II followed by anything
        name = re.sub(r'\s+(MD.*|DVM.*|DDS.*|PhD.*|II.*)', '', name)
        # Remove extra spaces
        name = re.sub(r'\s+', ' ', name).strip()
        return name

        data.loc[:, name_column] = data[name_column].apply(clean_name)
    return data
# Apply the function
data = remove_extra(data, 'Name')

# Display the cleaned DataFrame
data

Unnamed: 0                                    ID  \
0              0  1e407ff9-6255-489d-a0de-34135d4f74bd   
1              1  379f55b8-87d5-4739-a146-7400b78c24d1   
2              2  18261368-dfa1-47f0-afc6-bddf45926b07   
3              3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5   
4              4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90   
...          ...                                   ...   
9995        9995  610210a0-e344-4b79-95ce-ff6583d7d8ce   
9996        9996  eb66531c-901a-4da7-bc60-a4951886b5a8   
9997        9997  86a83f25-224d-498d-9901-c2e4a49ffbd2   
9998        9998  9cbba425-6af2-455f-8566-62402de79661   
9999        9999  ae6eee4d-ca8f-4932-96f1-487230875968   

                         Name  Age                              Email  \
0               Hunter Thomas   25                 xlopez@datahut.com   
1                Jeremy Irwin   90         jillianjenkins@datahut.com   
2     Jennifer Hammondquickly   66            jscottgreen@datahut.com   
3             Sydney Taylorso   39         luke56gonzalez@datahut.com   
4                   Julia Lee   71          figueroakayla@datahut.com   
...                       ...  ...                                ...   
9995          David Lambalone   70             jeremybanks@rogers.com   
9996              Karen Yoder   59  smithdaniel@williams-thompson.com   
9997        Anthony Rasmussen   57              nfranklin@datahut.com   
9998             Nicole Moore   82                 lmyers@datahut.com   
9999           Steven Gilbert   18                 cjones@datahut.com   

       Join Date    Salary    Department  
0     2022-03-31   88552.0         Sales  
1     2022-07-07  139227.0       Support  
2     2023-11-21   65550.0   Engineering  
3     2021-11-05  139932.0      SupportJ  
4     2022-03-31  143456.0     Marketing  
...          ...       ...           ...  
9995  2020-12-17   88472.0       Support  
9996  2023-12-12   71575.0       Support  
9997  05/12/1979  128456.0     Marketing  
9998  2020-10-28  128939.0  EngineeringV  
9999  2022-06-01  101447.0            HR  

[8902 rows x 8 columns]

## 8) Standardise Date Formats: 

In [423]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8902 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  8902 non-null   int64  
 1   ID          8902 non-null   object 
 2   Name        8902 non-null   object 
 3   Age         8902 non-null   int64  
 4   Email       8902 non-null   object 
 5   Join Date   8902 non-null   object 
 6   Salary      8902 non-null   float64
 7   Department  8902 non-null   object 
dtypes: float64(1), int64(2), object(5)
memory usage: 625.9+ KB


In [424]:
# Convert and format the Join Date(object) to Join Date(datetime) column
def convert_date(date):
    # Attempt to parse using different formats
    for fmt in ("%d/%m/%Y", "%Y/%m/%d", "%d-%m-%Y", "%Y-%m-%d"):
        try:
            return pd.to_datetime(date, format=fmt)  # Return as datetime
        except ValueError:
            continue
    return pd.NaT  # Return NaT if all formats fail

# Apply the conversion function to the 'Join Date' column

data['Join Date'] = data['Join Date'].apply(convert_date)


C:\Users\itsme\AppData\Local\Temp\ipykernel_31228\637017072.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['Join Date'] = data['Join Date'].apply(convert_date)


In [425]:
data.dtypes

Unnamed: 0             int64
ID                    object
Name                  object
Age                    int64
Email                 object
Join Date     datetime64[ns]
Salary               float64
Department            object
dtype: object

In [426]:
data

Unnamed: 0                                    ID  \
0              0  1e407ff9-6255-489d-a0de-34135d4f74bd   
1              1  379f55b8-87d5-4739-a146-7400b78c24d1   
2              2  18261368-dfa1-47f0-afc6-bddf45926b07   
3              3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5   
4              4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90   
...          ...                                   ...   
9995        9995  610210a0-e344-4b79-95ce-ff6583d7d8ce   
9996        9996  eb66531c-901a-4da7-bc60-a4951886b5a8   
9997        9997  86a83f25-224d-498d-9901-c2e4a49ffbd2   
9998        9998  9cbba425-6af2-455f-8566-62402de79661   
9999        9999  ae6eee4d-ca8f-4932-96f1-487230875968   

                         Name  Age                              Email  \
0               Hunter Thomas   25                 xlopez@datahut.com   
1                Jeremy Irwin   90         jillianjenkins@datahut.com   
2     Jennifer Hammondquickly   66            jscottgreen@datahut.com   
3             Sydney Taylorso   39         luke56gonzalez@datahut.com   
4                   Julia Lee   71          figueroakayla@datahut.com   
...                       ...  ...                                ...   
9995          David Lambalone   70             jeremybanks@rogers.com   
9996              Karen Yoder   59  smithdaniel@williams-thompson.com   
9997        Anthony Rasmussen   57              nfranklin@datahut.com   
9998             Nicole Moore   82                 lmyers@datahut.com   
9999           Steven Gilbert   18                 cjones@datahut.com   

      Join Date    Salary    Department  
0    2022-03-31   88552.0         Sales  
1    2022-07-07  139227.0       Support  
2    2023-11-21   65550.0   Engineering  
3    2021-11-05  139932.0      SupportJ  
4    2022-03-31  143456.0     Marketing  
...         ...       ...           ...  
9995 2020-12-17   88472.0       Support  
9996 2023-12-12   71575.0       Support  
9997 1979-12-05  128456.0     Marketing  
9998 2020-10-28  128939.0  EngineeringV  
9999 2022-06-01  101447.0            HR  

[8902 rows x 8 columns]

## 9) Correct Department Names:

In [427]:
# removed the extra noise came after each department name
data.loc[:,"Department"]=data["Department"].str.replace(r"sales.*","Sales",case=False,regex=True)
data.loc[:,"Department"]=data["Department"].str.replace(r"Support.*","Support",case=False,regex=True)
data.loc[:,"Department"]=data["Department"].str.replace(r"Engineering.*","Engineering",case=False,regex=True)
data.loc[:,"Department"]=data["Department"].str.replace(r"Marketing.*","Marketing",case=False,regex=True)
data.loc[:,"Department"]=data["Department"].str.replace(r"HR.*","HR",case=False,regex=True)
data

Unnamed: 0                                    ID  \
0              0  1e407ff9-6255-489d-a0de-34135d4f74bd   
1              1  379f55b8-87d5-4739-a146-7400b78c24d1   
2              2  18261368-dfa1-47f0-afc6-bddf45926b07   
3              3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5   
4              4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90   
...          ...                                   ...   
9995        9995  610210a0-e344-4b79-95ce-ff6583d7d8ce   
9996        9996  eb66531c-901a-4da7-bc60-a4951886b5a8   
9997        9997  86a83f25-224d-498d-9901-c2e4a49ffbd2   
9998        9998  9cbba425-6af2-455f-8566-62402de79661   
9999        9999  ae6eee4d-ca8f-4932-96f1-487230875968   

                         Name  Age                              Email  \
0               Hunter Thomas   25                 xlopez@datahut.com   
1                Jeremy Irwin   90         jillianjenkins@datahut.com   
2     Jennifer Hammondquickly   66            jscottgreen@datahut.com   
3             Sydney Taylorso   39         luke56gonzalez@datahut.com   
4                   Julia Lee   71          figueroakayla@datahut.com   
...                       ...  ...                                ...   
9995          David Lambalone   70             jeremybanks@rogers.com   
9996              Karen Yoder   59  smithdaniel@williams-thompson.com   
9997        Anthony Rasmussen   57              nfranklin@datahut.com   
9998             Nicole Moore   82                 lmyers@datahut.com   
9999           Steven Gilbert   18                 cjones@datahut.com   

      Join Date    Salary   Department  
0    2022-03-31   88552.0        Sales  
1    2022-07-07  139227.0      Support  
2    2023-11-21   65550.0  Engineering  
3    2021-11-05  139932.0      Support  
4    2022-03-31  143456.0    Marketing  
...         ...       ...          ...  
9995 2020-12-17   88472.0      Support  
9996 2023-12-12   71575.0      Support  
9997 1979-12-05  128456.0    Marketing  
9998 2020-10-28  128939.0  Engineering  
9999 2022-06-01  101447.0           HR  

[8902 rows x 8 columns]

In [ ]:
Noises after the Sales, Marketing,Engineering,Hr,Support are removed using string replacement method.

## 10. Handle Salary Noise:

In [428]:
# astype(int) convert salary(float) to salary(int)
data['Salary']=data['Salary'].astype(int)
data

C:\Users\itsme\AppData\Local\Temp\ipykernel_31228\2902414250.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['Salary']=data['Salary'].astype(int)


Unnamed: 0                                    ID  \
0              0  1e407ff9-6255-489d-a0de-34135d4f74bd   
1              1  379f55b8-87d5-4739-a146-7400b78c24d1   
2              2  18261368-dfa1-47f0-afc6-bddf45926b07   
3              3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5   
4              4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90   
...          ...                                   ...   
9995        9995  610210a0-e344-4b79-95ce-ff6583d7d8ce   
9996        9996  eb66531c-901a-4da7-bc60-a4951886b5a8   
9997        9997  86a83f25-224d-498d-9901-c2e4a49ffbd2   
9998        9998  9cbba425-6af2-455f-8566-62402de79661   
9999        9999  ae6eee4d-ca8f-4932-96f1-487230875968   

                         Name  Age                              Email  \
0               Hunter Thomas   25                 xlopez@datahut.com   
1                Jeremy Irwin   90         jillianjenkins@datahut.com   
2     Jennifer Hammondquickly   66            jscottgreen@datahut.com   
3             Sydney Taylorso   39         luke56gonzalez@datahut.com   
4                   Julia Lee   71          figueroakayla@datahut.com   
...                       ...  ...                                ...   
9995          David Lambalone   70             jeremybanks@rogers.com   
9996              Karen Yoder   59  smithdaniel@williams-thompson.com   
9997        Anthony Rasmussen   57              nfranklin@datahut.com   
9998             Nicole Moore   82                 lmyers@datahut.com   
9999           Steven Gilbert   18                 cjones@datahut.com   

      Join Date  Salary   Department  
0    2022-03-31   88552        Sales  
1    2022-07-07  139227      Support  
2    2023-11-21   65550  Engineering  
3    2021-11-05  139932      Support  
4    2022-03-31  143456    Marketing  
...         ...     ...          ...  
9995 2020-12-17   88472      Support  
9996 2023-12-12   71575      Support  
9997 1979-12-05  128456    Marketing  
9998 2020-10-28  128939  Engineering  
9999 2022-06-01  101447           HR  

[8902 rows x 8 columns]

In [ ]:
The Salary is formatted into integer.

In [430]:
data.dtypes

Unnamed: 0             int64
ID                    object
Name                  object
Age                    int64
Email                 object
Join Date     datetime64[ns]
Salary                 int64
Department            object
dtype: object

In [431]:
from IPython.display import FileLink

data.to_csv('cleaned_dataset.csv', index=False)
FileLink('cleaned_dataset.csv')

C:\Users\itsme\cleaned_dataset.csv

## Thankyou